File: core-java/src/main/java/com/kakao/actionbase/core/java/metadata/v3/TableDescriptor.java
package com.kakao.actionbase.core.java.metadata.v3;

import com.kakao.actionbase.core.java.edge.EdgePayload;
import com.kakao.actionbase.core.java.edge.EdgeState;
import com.kakao.actionbase.core.java.metadata.*;
import com.kakao.actionbase.core.java.metadata.ImmutableTableId;
import com.kakao.actionbase.core.java.metadata.TableId;
import com.kakao.actionbase.core.java.metadata.v3.common.MutationMode;
import com.kakao.actionbase.core.java.metadata.v3.common.Schema;
import com.kakao.actionbase.core.java.metadata.v3.common.SchemaType;
import com.kakao.actionbase.core.java.util.HashUtils;

import org.immutables.value.Value;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
  @JsonSubTypes.Type(value = VertexTableDescriptor.class, name = SchemaType.VERTEX_TYPE),
  @JsonSubTypes.Type(value = EdgeTableDescriptor.class, name = SchemaType.EDGE_TYPE),
})
public interface TableDescriptor<T extends Schema> extends Descriptor<TableId> {

  SchemaType type();

  String database();

  String table();

  String storage();

  T schema();

  MutationMode mode();

  @Override
  @JsonIgnore
  @Value.Derived
  @Value.Auxiliary
  default TableId id() {
    return ImmutableTableId.of(tenant(), database(), table());
  }

  @JsonIgnore
  @Value.Derived
  @Value.Auxiliary
  default StorageId storageId() {
    return ImmutableStorageId.of(tenant(), database(), storage());
  }

  @JsonIgnore
  @Value.Derived
  @Value.Auxiliary
  default int code() {
    String fullQualifiedName = database() + "." + table();
    return HashUtils.stringHash(fullQualifiedName);
  }

  @Override
  @JsonIgnore
  @Value.Auxiliary
  default EdgePayload toEdgePayload() {
    return Tables.toEdgePayload(this);
  }

  @Override
  @JsonIgnore
  @Value.Auxiliary
  default EdgeState toEdgeState() {
    return Tables.toEdgeState2(this);
  }
}

File: core/src/main/kotlin/com/kakao/actionbase/core/metadata/TableDescriptor.kt
package com.kakao.actionbase.core.metadata

import com.kakao.actionbase.core.Constants
import com.kakao.actionbase.core.metadata.common.ModelSchema
import com.kakao.actionbase.core.metadata.common.MutationMode

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeName

@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type",
)
@JsonSubTypes(
    JsonSubTypes.Type(value = TableDescriptor.Edge::class, name = "edge"),
    JsonSubTypes.Type(value = TableDescriptor.MultiEdge::class, name = "multiEdge"),
)
sealed class TableDescriptor<T : ModelSchema> : V3Descriptor<TableId> {
    abstract val database: String

    abstract val table: String

    abstract val schema: T

    abstract val mode: MutationMode

    abstract val storage: String

    @JsonTypeName("edge")
    data class Edge(
        override val tenant: String,
        override val database: String,
        override val table: String,
        override val schema: ModelSchema.Edge,
        override val mode: MutationMode,
        override val storage: String,
        override val active: Boolean = true,
        override val comment: String = Constants.DEFAULT_COMMENT,
        override val revision: Long = Constants.DEFAULT_REVISION,
        override val createdAt: Long = Constants.DEFAULT_CREATED_AT,
        override val createdBy: String = Constants.DEFAULT_CREATED_BY,
        override val updatedAt: Long = Constants.DEFAULT_UPDATED_AT,
        override val updatedBy: String = Constants.DEFAULT_UPDATED_BY,
    ) : TableDescriptor<ModelSchema.Edge>() {
        @JsonIgnore
        override val id: TableId = TableId(tenant, database, table)
    }

    @JsonTypeName("multiEdge")
    data class MultiEdge(
        override val tenant: String,
        override val database: String,
        override val table: String,
        override val schema: ModelSchema.MultiEdge,
        override val mode: MutationMode,
        override val storage: String,
        override val active: Boolean = true,
        override val comment: String = Constants.DEFAULT_COMMENT,
        override val revision: Long = Constants.DEFAULT_REVISION,
        override val createdAt: Long = Constants.DEFAULT_CREATED_AT,
        override val createdBy: String = Constants.DEFAULT_CREATED_BY,
        override val updatedAt: Long = Constants.DEFAULT_UPDATED_AT,
        override val updatedBy: String = Constants.DEFAULT_UPDATED_BY,
    ) : TableDescriptor<ModelSchema.MultiEdge>() {
        @JsonIgnore
        override val id: TableId = TableId(tenant, database, table)
    }
}
